### ISSUE TITLE
install.sh can't easily be run headlessly

### ISSUE THREAD

**Issue by Xyene** [ocaml/opam#4136]
In `install.sh`, linked from [How to install opam](https://opam.ocaml.org/doc/Install.html), the install path is [unconditionally queried from the user](https://github.com/ocaml/opam/blob/master/shell/install.sh#L206).

This makes it less than ideal for scripting. What I've resorted to doing is:

```bash
echo | sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)
```

...which is a bit of a hack.

Another small issue is that the script is [marked as runnable by `/bin/sh`](https://github.com/ocaml/opam/blob/master/shell/install.sh#L1), but the process substitution syntax from the documentation (`sh <(...)`) is not particularly portable outside of `/bin/bash`. This stumped me for a bit while integrating this step in a Docker build, where the default shell is not Bash. 

(I did notice there was a nudge to use a prebuilt Docker image, but that's not always possible -- in my case, I'm installing OCaml alongside several other runtimes, and can't base off of `ocaml/opam2`.)

I'm happy to submit PRs for both these issues if they should, in fact, be fixed.

**Comment by rjbou** [comment]
The `install.sh` script is not intend to be used in scripts. Why do you need to use it in a docker build ?
On the bash part, yes, we took a _mosts_ generic one-line install instruction, but there is always the possibility for user to download the script and launch if doesn't work.

Related to #3500

**Comment by Xyene** [comment]
Thanks for the quick reply! In my case, I can't base off of `ocaml/opam2`, since I'm basing off of something else. I initially installed `opam` via my distribution's (Debian's) package manager, but ran into https://github.com/ocaml/opam/issues/3827.

Admittedly, there is a workaround in that thread, but I did not feel comfortable inserting flags that I don't fully understand the repercussions of into a large Dockerfile that I likely won't touch for _N_ months (hopefully for some very large _N_). I also noticed that the latest binaries from GitHub releases don't suffer from that problem.

I could download the archive from GitHub directly, but then I'd end up recreating a large part of `install.sh` myself, and there's a nontrivial amount of code for useful things like signature validation that are already handled by the script. By that point, even an `echo | ... install.sh` hack seemed more preferable to the alternative.

(Note that I totally missed the CI note until after I opened this issue; I guess I've been trained by other runtime installers that the presence of a script ⇒ the ability to copy/paste that script into a build, even if it's not the case here.)

Regarding the process substitution issue, I (via Google) happened across an [older version of the document](https://opam.ocaml.org/doc/1.2/Install.html) that had a more traditional pipe approach. I don't have particularly strong feelings on the matter, but that seems a less surprising method (conditioned by the installer scripts for [Rust](https://rustup.rs/), and [Groovy](https://groovy-lang.org/install.html), Scala, Kotlin etc. via sdkman).

Circling back, I can totally understand the stance of not wanting to maintain `install.sh` as an automatable interface. At least in this case, though, the alternatives that I've found aren't great.

**Comment by rjbou** [comment]
I understand. We could add an argument to the `install.sh` script for the download-only actions: os resolution and sha check. Signatures are available on github release page, but are not checked by the script.
We should also add a warning on the old script, or even delete it (who still uses opam1.2?). It was kept only for legacy reasons.

**Comment by rjbou** [comment]
Fixed by #4306, closing

**Comment by Xyene** [comment]
Thanks @rjbou!
